File: tests/test_dff.cpp
// Copyright cocotb contributors
// Licensed under the Revised BSD License, see LICENSE for details.
// SPDX-License-Identifier: BSD-3-Clause

#include <cocotb.h>

using namespace cocotb;

COCOTB_TEST(test_dff) // registers the test
task<> test_dff(Dut &dut) {
    cocotb::log.info("Starting test_dff coroutine");

    // Set initial input value to prevent it from floating
    dut["d"].value = 0;

    // Create a 10us period clock driver on port `clk`
    auto clock = Clock(dut, 10, unit::us);

    // Start the clock
    auto clk = start_soon(clock);

    // Synchronize with the clock. This will register the initial `d` value
    co_await RisingEdge(dut["clk"]);

    auto expected_val = 0; // Matches initial input value
    for (int i = 0; i < 10; ++i) {
        auto val = rand() % 2;
        dut["d"].value = val; // Assign the random value val to the input port d
        co_await RisingEdge(dut["clk"]);
        auto q_val = dut["q"].value;
        assert(expected_val == q_val, std::format("output q was incorrect on the {} th cycle", i));
        expected_val = val; // Save random value for next RisingEdge
    }

    co_await RisingEdge(dut["clk"]);
    assert(expected_val == dut["q"].value, "output q was incorrect on the last cycle");

    cocotb::log.info("test_dff completed successfully");
}

task<> Wait(Dut &dut, int time) {
    co_await Timer(time);
    dut["clk"].value = 0;
    co_await Timer(time);
    dut["d"].value = 1;
    co_await Timer(time);
    dut["clk"].value = 1;
    co_await Timer(time);
    dut["d"].value = 0;
}

COCOTB_TEST(test_dff_post)
task<> test_dff_post(Dut &dut) {
    cocotb::log.info("Starting test_dff_post coroutine");

    dut["d"].value = 0;
    dut["clk"].value = 0;

    auto d_int = int32_t(dut["d"].value);
    cocotb::log.info("d_int: {}", d_int);

    uint32_t d_uint = dut["d"].value;
    cocotb::log.info("d_uint: {}", d_uint);

    co_await Wait(dut, 20);
    co_await Timer(10);

    assert(1 == dut["q"].value, "1: output q was incorrect");

    co_await Timer(10);

    dut["clk"].value = 0;
    dut["d"].value = 0;
    co_await Timer(10);
    dut["clk"].value = 1;
    co_await Timer(10);
    dut["d"].value = 0;

    assert(0 == dut["q"].value, "2: output q was incorrect");

    co_await Wait(dut, 100);
    cocotb::log.info("Create wait coroutine");
    auto wait = start_soon(Wait(dut, 100));
    cocotb::log.info("Awaiting wait coroutine");
    co_await wait;
    cocotb::log.info("Awaited wait coroutine");

    co_await Timer(10);
    assert(1 == dut["q"].value, "3: output q was incorrect");

    co_await Timer(10, unit::us);

    auto sim_time = get_sim_time();
    cocotb::log.info("sim time: {}", sim_time);

    co_await Timer(10);
    assert(sim_time + 10 == get_sim_time());

    co_await Timer(10, unit::us);
    assert(sim_time + 10 + 10000 == get_sim_time());

    cocotb::log.info("Completed test_dff_post successfully");
}

File: tests/test_dff.py
# This file is public domain, it can be freely copied without restrictions.
# SPDX-License-Identifier: CC0-1.0
from __future__ import annotations

import os
import random
from pathlib import Path

import cocotb
from cocotb.clock import Clock
from cocotb.triggers import RisingEdge, Timer
from cocotb_tools.runner import get_runner


@cocotb.test()
async def dff_test(dut):
    """Test that d propagates to q"""

    # Set initial input value to prevent it from floating
    dut.d.value = 0

    # Create a 10us period clock driver on port `clk`
    clock = Clock(dut.clk, 10, unit="us")
    # Start the clock. Start it low to avoid issues on the first RisingEdge
    clock.start(start_high=False)

    # Synchronize with the clock. This will register the initial `d` value
    await RisingEdge(dut.clk)

    expected_val = 0  # Matches initial input value
    for i in range(10):
        val = random.randint(0, 1)
        dut.d.value = val  # Assign the random value val to the input port d
        await RisingEdge(dut.clk)
        assert dut.q.value == expected_val, f"output q was incorrect on the {i}th cycle"
        expected_val = val  # Save random value for next RisingEdge

    # Check the final input on the next clock
    await RisingEdge(dut.clk)
    assert dut.q.value == expected_val, "output q was incorrect on the last cycle"


async def Wait(dut, time):
    await Timer(time)
    dut.d.value = 1
    dut.clk.value = 1


@cocotb.test()
async def test_dff_post(dut):
    dut.d.value = 0
    dut.clk.value = 0

    wait = cocotb.start_soon(Wait(dut, 100))

    await Wait(dut, 10)

    await wait

    await Timer(10)
    assert dut.d.value == 1, "output d was incorrect"


def test_dff_runner():
    sim = os.getenv("SIM", "icarus")

    proj_path = Path(__file__).resolve().parent

    sources = [proj_path / "dff.sv"]

    runner = get_runner(sim)
    runner.build(
        sources=sources,
        hdl_toplevel="dff",
        always=True,
    )

    runner.test(hdl_toplevel="dff", test_module="test_dff,")


if __name__ == "__main__":
    test_dff_runner()
